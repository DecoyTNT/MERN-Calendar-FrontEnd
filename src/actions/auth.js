import { types } from './../types/types';
import { clientAxios } from './../config/axios';
import Swal from 'sweetalert2';
import { tokenAuth } from './../config/token';

export const startLogin = (email, password) => {
    return async dispatch => {
        try {
            const resp = await clientAxios.post('/auth', { email, password });
            if (resp.data.ok) {
                localStorage.setItem('token', resp.data.token);
                localStorage.setItem('token-init-date', new Date().getTime());

                const { user } = resp.data;
                dispatch(login(user));
            }
        } catch (error) {
            Swal.fire(
                'Error',
                error.response.data.msg,
                'error'
            );
        }
    }
}

export const startRegister = (name, email, password) => {
    return async dispatch => {
        try {
            const resp = await clientAxios.post('/auth/new', { name, email, password });
            if (resp.data.ok) {
                localStorage.setItem('token', resp.data.token);
                localStorage.setItem('token-init-date', new Date().getTime());

                const { user } = resp.data;
                dispatch(login(user));
            }
        } catch (error) {
            Swal.fire(
                'Error',
                error.response.data.msg,
                'error'
            );
        }
    }
}

export const startChecking = () => {
    return async dispatch => {
        try {
            const token = localStorage.getItem('token');
            if (!token) {
                return dispatch(checkingFinish());
            }
            tokenAuth(token);
            const resp = await clientAxios.get('/auth/renew');
            if (resp.data.ok) {
                localStorage.setItem('token', resp.data.token);
                localStorage.setItem('token-init-date', new Date().getTime());

                const { user } = resp.data;
                dispatch(login(user));
            }

        } catch (error) {
            dispatch(checkingFinish());
            localStorage.clear();
        }
    }
}

const checkingFinish = () => ({
    type: types.authCheckingFinish
});

const login = user => ({
    type: types.authLogin,
    payload: user
});

export const startLogout = () => {
    return dispatch => {
        localStorage.clear();
        dispatch(logout());
    }
}

const logout = () => ({
    type: types.authLogout
});