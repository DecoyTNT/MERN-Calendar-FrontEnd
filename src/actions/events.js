import { types } from './../types/types';
import { clientAxios } from './../config/axios';
import { prepareEvents } from './../helpers/prepareEvents';
import { tokenAuth } from './../config/token';
import Swal from 'sweetalert2';

export const eventStartAddNew = event => {
    return async (dispatch, getState) => {

        const { uid, name } = getState().auth;

        try {
            const resp = await clientAxios.post('/events', event);
            if (resp.data.ok) {
                event.id = resp.data.event.id;
                event.user = {
                    _id: uid,
                    name
                }
                dispatch(eventAddNew(event));
            }
        } catch (error) {
            Swal.fire(
                'Error',
                error.response.data.msg,
                'error'
            );
        }
    }
}

const eventAddNew = event => ({
    type: types.eventAddNew,
    payload: event
});

export const eventSetActive = event => ({
    type: types.eventSetActive,
    payload: event
});

export const eventClearActiveEvent = () => ({
    type: types.eventClearActiveEvent
});

export const eventStartUpdate = (event) => {
    return async dispatch => {
        try {
            const resp = await clientAxios.put(`/events/${event.id}`, event);
            if (resp.data.ok) {
                dispatch(eventUpdated(event));
            }
        } catch (error) {
            Swal.fire(
                'Error',
                error.response.data.msg,
                'error'
            );
        }
    }
}

const eventUpdated = event => ({
    type: types.eventUpdated,
    payload: event
});

export const eventStartDelete = () => {
    return async (dispatch, getState) => {

        const { activeEvent } = getState().calendar;

        try {
            Swal.fire({
                title: '¿Estás seguro?',
                text: "Ya no se podra recuperar una vez eliminado",
                icon: 'warning',
                showCancelButton: true,
                confirmButtonColor: '#3085d6',
                cancelButtonColor: '#d33',
                confirmButtonText: 'Si, Eliminar!',
                cancelButtonText: 'No, Cancelar',
            }).then(async (result) => {
                if (result.isConfirmed) {
                    const resp = await clientAxios.delete(`/events/${activeEvent.id}`);
                    if (resp.data.ok) {
                        dispatch(eventDeleted());
                    }
                    Swal.fire(
                        'Eliminado!',
                        'El evento fue eliminado.',
                        'success'
                    )
                }
            })
        } catch (error) {
            Swal.fire(
                'Error',
                error.response.data.msg,
                'error'
            );
        }
    }
}

const eventDeleted = () => ({
    type: types.eventDeleted
});

export const eventSetStartEnd = event => ({
    type: types.eventSetStartEnd,
    payload: event
});

export const eventsStartLoading = () => {
    return async dispatch => {
        try {
            const token = localStorage.getItem('token');
            tokenAuth(token);
            const resp = await clientAxios.get('/events');
            if (resp.data.ok) {
                const events = prepareEvents(resp.data.events);
                dispatch(eventsLoaded(events));
            }
        } catch (error) {
            console.log(error);
        }
    }
}

export const eventsUserStartLoading = () => {
    return async dispatch => {
        try {
            const token = localStorage.getItem('token');
            tokenAuth(token);
            const resp = await clientAxios.get('/events/user');
            if (resp.data.ok) {
                const events = prepareEvents(resp.data.events);
                dispatch(eventsLoaded(events));
            }
        } catch (error) {
            console.log(error);
        }
    }
}

const eventsLoaded = events => ({
    type: types.eventsLoaded,
    payload: events
});

export const eventCleanAll = () => ({
    type: types.eventCleanAll
});