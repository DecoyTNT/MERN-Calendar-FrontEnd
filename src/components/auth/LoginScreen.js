import React from 'react';
import { useFormik } from 'formik';
import * as Yup from 'yup';
import { Link } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import { startLogin, startChecking } from './../../actions/auth';
import './auth.css';

const LoginScreen = () => {

    const dispatch = useDispatch();

    const formik = useFormik({
        initialValues: {
            email: '',
            password: ''
        },
        validationSchema: Yup.object({
            email: Yup.string()
                .required('El correo es obligatorio')
                .email('El email no es valido'),
            password: Yup.string()
                .required('El password es obligatorio')
                .min(6, 'El password debe contener al menos 6 carÃ¡cteres')
        }),
        onSubmit: () => {
            dispatch(startLogin(email, password));
            dispatch(startChecking());
        }
    });

    const { email, password } = formik.values;

    return (
        <div className="container login-container">
            <div className="col-md-6 login-form-1">
                <h3>Ingreso</h3>
                <form
                    onSubmit={formik.handleSubmit}
                >
                    <div className="form-group">
                        <input
                            type="email"
                            className="form-control"
                            placeholder="Correo"
                            name="email"
                            value={email}
                            onChange={formik.handleChange}
                        />
                    </div>
                    <div className="form-group">
                        <input
                            type="password"
                            className="form-control"
                            placeholder="Password"
                            name="password"
                            value={password}
                            onChange={formik.handleChange}
                        />
                    </div>
                    <div className="form-group">
                        <input
                            type="submit"
                            className="btnSubmit"
                            value="Login"
                        />
                    </div>
                    <Link to="/register">Crear nueva cuenta</Link>
                </form>
            </div>
        </div>
    )
}

export default LoginScreen;