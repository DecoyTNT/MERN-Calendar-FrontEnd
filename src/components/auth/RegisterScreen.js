import React from 'react';
import { Link } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import { useFormik } from 'formik';
import * as Yup from 'yup';
import './auth.css';
import { startRegister } from '../../actions/auth';

const RegisterScreen = () => {

    const dispatch = useDispatch();

    const formik = useFormik({
        initialValues: {
            name: '',
            email: '',
            password: '',
            password2: ''
        },
        validationSchema: Yup.object({
            name: Yup.string()
                .required('El nombre es obligatorio'),
            email: Yup.string()
                .required('El correo es obligatorio')
                .email('El email no es valido'),
            password: Yup.string()
                .required('El password es obligatorio')
                .min(6, 'El password debe contener al menos 6 carÃ¡cteres'),
            password2: Yup.string()
                .test('passwords-match', 'Los passwords deben ser iguales', function (value) {
                    return this.parent.password === value;
                })
        }),
        validateOnChange: false,
        onSubmit: () => {
            dispatch(startRegister(name, email, password));
        }
    });

    const { name, email, password, password2 } = formik.values;

    return (
        <div className="container login-container">
            <div className="col-md-6 login-form-2">
                <h3>Registro</h3>
                <form
                    onSubmit={formik.handleSubmit}
                >
                    {formik.errors.name &&
                        (
                            <div className="alert bg-danger" role="alert">
                                <p className="font-weight-bold text-white">{formik.errors.name}</p>
                            </div>
                        )}
                    <div className="form-group">
                        <input
                            type="text"
                            className="form-control"
                            placeholder="Nombre"
                            name="name"
                            value={name}
                            onChange={formik.handleChange}
                        />
                    </div>
                    {formik.errors.email &&
                        (
                            <div className="alert bg-danger" role="alert">
                                <p className="font-weight-bold text-white">{formik.errors.email}</p>
                            </div>
                        )}
                    <div className="form-group">
                        <input
                            type="email"
                            className="form-control"
                            placeholder="Correo"
                            name="email"
                            value={email}
                            onChange={formik.handleChange}
                        />
                    </div>
                    {formik.errors.password &&
                        (
                            <div className="alert bg-danger" role="alert">
                                <p className="font-weight-bold text-white">{formik.errors.password}</p>
                            </div>
                        )}
                    <div className="form-group">
                        <input
                            type="password"
                            className="form-control"
                            placeholder="Password"
                            name="password"
                            value={password}
                            onChange={formik.handleChange}
                        />
                    </div>
                    {formik.errors.password2 &&
                        (
                            <div className="alert bg-danger" role="alert">
                                <p className="font-weight-bold text-white">{formik.errors.password2}</p>
                            </div>
                        )}
                    <div className="form-group">
                        <input
                            type="password"
                            className="form-control"
                            placeholder="Repite el password"
                            name="password2"
                            value={password2}
                            onChange={formik.handleChange}
                        />
                    </div>

                    <div className="form-group">
                        <input
                            type="submit"
                            className="btnSubmit"
                            value="Crear cuenta" />
                    </div>
                    <Link to="/login">Ya tengo cuenta</Link>
                </form>
            </div>
        </div>
    )
}

export default RegisterScreen;
