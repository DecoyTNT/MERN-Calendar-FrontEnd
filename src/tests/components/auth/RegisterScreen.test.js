import { BrowserRouter as Router } from 'react-router-dom';
import { act } from 'react-dom/test-utils';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import '@testing-library/jest-dom';
import RegisterScreen from './../../../components/auth/RegisterScreen';
import { startRegister } from './../../../actions/auth';

jest.mock('../../../actions/auth', () => ({
    startRegister: jest.fn()
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {};
const store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={store}>
        <Router>
            <RegisterScreen />
        </Router>
    </Provider>
);

describe('Pruebas en <RegisterScreen />', () => {

    test('Debe mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('No hay registro si las contraseÃ±as no son iguales', async () => {
        wrapper.find('input[name="name"]').simulate('change', {
            target: {
                name: 'name',
                value: 'TEST'
            }
        });

        wrapper.find('input[name="email"]').simulate('change', {
            target: {
                name: 'email',
                value: 'test@test.com'
            }
        });

        wrapper.find('input[name="password"]').simulate('change', {
            target: {
                name: 'password',
                value: '123456'
            }
        });

        wrapper.find('input[name="password2"]').simulate('change', {
            target: {
                name: 'password2',
                value: '1234567'
            }
        });

        await act(async () => {
            wrapper.find('form').prop('onSubmit')();
            await new Promise(resolve => setTimeout(resolve));
            wrapper.update();
        });
        expect(startRegister).not.toHaveBeenCalled();
        expect(wrapper.find('p').text()).toBe('Los passwords deben ser iguales');

    });

    test('Debe de hacer el registro', async () => {
        wrapper.find('input[name="name"]').simulate('change', {
            target: {
                name: 'name',
                value: 'TEST'
            }
        });

        wrapper.find('input[name="email"]').simulate('change', {
            target: {
                name: 'email',
                value: 'test@test.com'
            }
        });

        wrapper.find('input[name="password"]').simulate('change', {
            target: {
                name: 'password',
                value: '123456'
            }
        });

        wrapper.find('input[name="password2"]').simulate('change', {
            target: {
                name: 'password2',
                value: '123456'
            }
        });

        await act(async () => {
            wrapper.find('form').prop('onSubmit')();
        });

        expect(startRegister).toHaveBeenCalled();
        expect(startRegister).toHaveBeenCalledWith('TEST', 'test@test.com', '123456');
    });

});