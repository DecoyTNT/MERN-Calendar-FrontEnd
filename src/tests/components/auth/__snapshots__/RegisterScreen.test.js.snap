// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <RegisterScreen /> Debe mostrarse correctamente 1`] = `
<div
  className="container login-container"
>
  <div
    className="col-md-6 login-form-2"
  >
    <h3>
      Registro
    </h3>
    <form
      onSubmit={[Function]}
    >
      <div
        className="form-group"
      >
        <input
          className="form-control"
          name="name"
          onChange={[Function]}
          placeholder="Introduce tu nombre"
          type="text"
          value=""
        />
      </div>
      <div
        className="form-group"
      >
        <input
          className="form-control"
          name="email"
          onChange={[Function]}
          placeholder="Introduce tu correo"
          type="email"
          value=""
        />
      </div>
      <div
        className="form-group"
      >
        <input
          className="form-control"
          name="password"
          onChange={[Function]}
          placeholder="Introduce tu password"
          type="password"
          value=""
        />
      </div>
      <div
        className="form-group"
      >
        <input
          className="form-control"
          name="password2"
          onChange={[Function]}
          placeholder="Vuelve a introducir tu password"
          type="password"
          value=""
        />
      </div>
      <div
        className="form-group"
      >
        <input
          className="btnSubmit"
          type="submit"
          value="Crear cuenta"
        />
      </div>
      <a
        href="/login"
        onClick={[Function]}
      >
        Ya tengo cuenta
      </a>
    </form>
  </div>
</div>
`;
